Core/Debugger/PythonScripting.cpp

```diff
@@ -0,0 +1,348 @@
+g_on_breakpoint_callback = NULL;
+
+static PyObject *
+g_on_frame_callback = NULL;
+
+static PyObject *
+g_python_globals = NULL;
+
+void
+ppsspp_py_on_breakpoint_callback(u32 addr, void *user_data)
+{
+	if (g_on_breakpoint_callback && PyCallable_Check(g_on_breakpoint_callback)) {
+		PyObject *result = PyObject_CallFunction(g_on_breakpoint_callback, "I", addr);
+		if (result == NULL) {
+			NOTICE_LOG(SYSTEM, "Error calling Python function, see log");
+			PyErr_PrintEx(1);
+			PyErr_Clear();
+		} else {
+			// Unused result for now
+			Py_DECREF(result);
+		}
+	} else {
+		NOTICE_LOG(SYSTEM, "Got a breakpoint callback, but no callable object set");
+	}
+}
+
+PyObject *
+ppsspp_py_add_breakpoint(PyObject *self, PyObject *o)
+{
+	u32 addr = PyLong_AsLong(o);
+
+	CBreakPoints::AddBreakPoint(addr, false);
+	CBreakPoints::ChangeBreakPoint(addr, BREAK_ACTION_CALLBACK);
+
+	CBreakPoints::SetBreakpointCallback(ppsspp_py_on_breakpoint_callback, NULL);
+
+	Py_RETURN_NONE;
+}
+
+PyObject *
+ppsspp_py_remove_breakpoint(PyObject *self, PyObject *o)
+{
+	u32 addr = PyLong_AsLong(o);
+
+	CBreakPoints::RemoveBreakPoint(addr);
+
+	Py_RETURN_NONE;
+}
+
+PyObject *
+ppsspp_py_get_breakpoints(PyObject *self)
+{
+	PyObject *result = PyList_New(0);
+
+	for (const BreakPoint &bp: CBreakPoints::GetBreakpoints()) {
+		PyList_Append(result, PyLong_FromLong(bp.addr));
+	}
+
+	return result;
+}
+
+PyObject *
+ppsspp_py_on_breakpoint(PyObject *self, PyObject *callback)
+{
+	Py_INCREF(callback);
+
+	if (g_on_breakpoint_callback != NULL) {
+		Py_DECREF(g_on_breakpoint_callback);
+	}
+
+	g_on_breakpoint_callback = callback;
+
+	Py_RETURN_NONE;
+}
+
+PyObject *
+ppsspp_py_on_frame(PyObject *self, PyObject *callback)
+{
+	Py_INCREF(callback);
+
+	if (g_on_frame_callback != NULL) {
+		Py_DECREF(g_on_frame_callback);
+	}
+
+	g_on_frame_callback = callback;
+
+	Py_RETURN_NONE;
+}
+
+PyObject *
+ppsspp_py_get_reg_value(PyObject *self, PyObject *reg)
+{
+	return PyLong_FromLong(currentDebugMIPS->GetRegValue(0, PyLong_AsLong(reg)));
+}
+
+PyObject *
+ppsspp_py_read_memory_u8(PyObject *self, PyObject *addr)
+{
+	return PyLong_FromLong(Memory::ReadUnchecked_U8(PyLong_AsLong(addr)));
+}
+
+PyObject *
+ppsspp_py_write_memory_u8(PyObject *self, PyObject *args)
+{
+	unsigned long addr;
+	unsigned long value;
+	if (!PyArg_ParseTuple(args, "kk", &addr, &value)) {
+	    return NULL;
+	}
+
+	Memory::WriteUnchecked_U8(addr, value & 0xFF);
+
+	Py_RETURN_NONE;
+}
+
+PyObject *
+ppsspp_py_get_game_id(PyObject *self)
+{
+	std::string gamePath = PSP_CoreParameter().fileToStart;
+	std::shared_ptr<GameInfo> ginfo = g_gameInfoCache->GetInfo(nullptr, gamePath, 0);
+
+	return PyUnicode_FromString(ginfo->id.c_str());
+}
+
+PyObject *
+ppsspp_py_get_memory_pointer(PyObject *self)
+{
+	return PyLong_FromVoidPtr(Memory::base);
+}
+
+PyObject *
+ppsspp_py_get_ram_memory_view(PyObject *self)
+{
+	return PyMemoryView_FromMemory((char *)Memory::base + 0x08000000, 32ul * 1024ul * 1024ul, PyBUF_READ);
```

Comment:
Technically, should check the actual memory size (might be larger than 32 MB for some homebrew and PS3 remasters.)

-[Unknown]

Reply:
Is there a way to query the size from the PPSSPP Core?

Reply:
I suppose `Memory::g_MemorySize` is the best way.  You might also use `PSP_GetKernelMemoryBase()` and `PSP_GetUserMemoryEnd() - PSP_GetKernelMemoryBase()`

-[Unknown].